### ISSUE TITLE
aws_api_gateway_deployment redeploy causing inconsistant plan

### ISSUE THREAD

**Issue by njdart** [hashicorp/terraform-provider-aws#11469]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

```
Terraform v0.12.18
+ provider.aws v2.41.0
+ provider.random v2.2.0
+ provider.template v2.1.2
```

### Affected Resource(s)

* aws_api_gateway_deployment
* (aws_cloudfront_distribution)
* (aws_cloudwatch_dashboard)

### Terraform Configuration Files

**module.backend**
```hcl
resource "aws_api_gateway_rest_api" "lambda_api" {
  name = "Lambda API ${var.name}"
  body = data.template_file.api_swagger.rendered

  endpoint_configuration {
    types = ["REGIONAL"]
  }

  tags = {
    Name    = var.name
    Region  = var.region
    Profile = var.profile
  }
}

resource "aws_api_gateway_deployment" "api" {
  rest_api_id = aws_api_gateway_rest_api.lambda_api.id
  # The stage name will form part of the URL
  stage_name = "api"

  variables = {
    # We have to use this to force a re-deploy of the stage when the api swagger changes
    version = md5(data.template_file.api_swagger.rendered)
  }
}
```

**module.domainName**
```hcl
resource "aws_cloudfront_distribution" "main" {
  enabled             = true
  default_root_object = "index.html"

  aliases = [
    var.r53_domain_name
  ]

  origin {
    # Origin for our frontend cloudfront/S3
    domain_name = replace(replace(var.frontend_url, "/^https?:\\/\\//", ""), "/\\/.*$/", "")
    origin_id   = "${var.name} - Frontend Assets"
    custom_origin_config {
      http_port              = 80
      https_port             = 443
      origin_protocol_policy = "https-only"
      origin_ssl_protocols   = ["TLSv1.2"]
    }
  }

  default_cache_behavior {
    allowed_methods  = ["HEAD", "GET", "OPTIONS"]
    cached_methods   = ["HEAD", "GET", "OPTIONS"]
    target_origin_id = "${var.name} - Frontend Assets"
    compress         = true
    min_ttl          = 0
    default_ttl      = 0
    max_ttl          = 0

    forwarded_values {
      query_string = false
      cookies {
        forward = "none"
      }
    }

    viewer_protocol_policy = "redirect-to-https"
  }

  origin {
    # Origin for our backend API Gateway/Lambda
    domain_name = replace(replace(var.api_url, "/^https?:\\/\\//", ""), "/\\/.*$/", "")
    origin_id   = "${var.name} - Backend API"
    origin_path = ""

    custom_origin_config {
      http_port              = 80
      https_port             = 443
      origin_protocol_policy = "https-only"
      origin_ssl_protocols   = ["TLSv1.2"]
    }
  }

  ordered_cache_behavior {
    allowed_methods  = ["HEAD", "DELETE", "POST", "GET", "OPTIONS", "PUT", "PATCH"]
    cached_methods   = ["HEAD", "GET", "OPTIONS"]
    target_origin_id = "${var.name} - Backend API"
    path_pattern     = "/api*"
    compress         = true
    min_ttl          = 0
    default_ttl      = 0
    max_ttl          = 0

    forwarded_values {
      query_string = true
      cookies {
        forward           = "whitelist"
        whitelisted_names = ["authorisation"]
      }
    }

    viewer_protocol_policy = "redirect-to-https"
  }

  price_class = var.cloudfront_price_class

  restrictions {
    geo_restriction {
      restriction_type = "none"
    }
  }

  viewer_certificate {
    acm_certificate_arn      = var.r53_cert_arn
    minimum_protocol_version = "TLSv1.1_2016"
    ssl_support_method       = "sni-only"
  }

  tags = {
    Name    = var.name
    Region  = var.region
    Profile = var.profile
  }
}
```

### Debug Output

https://gist.github.com/njdart/fd7984e6a724327645a000e28141c9f4

### Panic Output

```
Error: Provider produced inconsistent final plan

When expanding the plan for module.domainName.aws_cloudfront_distribution.main
to include new values learned so far during apply, provider
"registry.terraform.io/-/aws" produced an invalid new value for .origin:
planned set element
cty.ObjectVal(map[string]cty.Value{"custom_header":cty.SetValEmpty(cty.Object(map[string]cty.Type{"name":cty.String,
"value":cty.String})),
"custom_origin_config":cty.ListVal([]cty.Value{cty.ObjectVal(map[string]cty.Value{"http_port":cty.NumberIntVal(80),
"https_port":cty.NumberIntVal(443),
"origin_keepalive_timeout":cty.NumberIntVal(5),
"origin_protocol_policy":cty.StringVal("https-only"),
"origin_read_timeout":cty.NumberIntVal(30),
"origin_ssl_protocols":cty.SetVal([]cty.Value{cty.StringVal("TLSv1.2")})})}),
"domain_name":cty.StringVal("<redacted>.cloudfront.net"),
"origin_id":cty.StringVal("nic-test - Frontend Assets"),
"origin_path":cty.NullVal(cty.String),
"s3_origin_config":cty.ListValEmpty(cty.Object(map[string]cty.Type{"origin_access_identity":cty.String}))})
does not correlate with any element in actual.
```

### Expected Behavior

Terraform should update the api gateway, keeping the current invocation url, name, id, etc the same. Alternatively, Terraform should destroy and recreate the deployment, and dependencies of the deployment should receive the correct deployment invocation url, name, id, etc.

### Actual Behavior

Terraform is destroying the existing API gateway deployment and creating a new one. This is causing the cloudfront distrubution and cloudwatch dashboard to be modified. The values received for the api gateway seem to be the `name` specified in the swagger file. I'm not sure why it's modifying both origins in the cloudfront distribution. I can't see any difference between the old and new versions

### Steps to Reproduce

Modifying the swagger definitions used in the API cause this issue to occur. Alternatively, changing `version = md5(...)` to `version = timestamp()` causes this issue to occur on each apply. If the swagger (or version variable) is unchanged, this issue does not occur. 

1. `terraform apply`

### Important Factoids

When I follow the [example api_gateway_deployment](https://www.terraform.io/docs/providers/aws/r/api_gateway_deployment.html) on the terraform docs, this does not occur.

### References

* I was following [this guide](https://aws.amazon.com/premiumsupport/knowledge-center/api-gateway-cloudfront-distribution/) to using a custom cloudfront distributions for an API gateway 
* I was following hashicorp/terraform#6613 for instructions on re-deploying an API Gateway on change

**Comment by njdart** [comment]
[I've made a minimal set of TF that reproduces part of the problem](https://gist.github.com/njdart/d4971d0744d5a15c1dfc4ed753328a85) (this example eventually becomes consistent, my real TF never does).

First off, i noticed that the first time you change the swagger and apply, the name of the `aws_api_gateway_rest_api` is the one provided in terraform;
```
        name                     = "Test API Gateway"
```

If you make a second apply, terraform will tell you that the name and description of the gateway that was made is actually the properties defined in the `info` section in the swagger, and needs to be changed back:
```
      - description              = "Test" -> null
        execution_arn            = "arn:aws:execute-api:eu-west-2:144204619291:ght35dsbkf"
        id                       = "ght35dsbkf"
        minimum_compression_size = -1
      ~ name                     = "Test - my awesome deployment" -> "Test API Gateway"
```

I changed my swagger to remove the `description` property, and changed the `title` to the same value as in my `aws_api_gateway_rest_api.main` resource. This solved this part.

I made this change in my production code, and i'm still getting modifications to the cloudfront distribution:

```
      - origin {
          - domain_name = "<redacted>.execute-api.eu-west-2.amazonaws.com" -> null
          - origin_id   = "nic-test - Backend API" -> null

          - custom_origin_config {
              - http_port                = 80 -> null
              - https_port               = 443 -> null
              - origin_keepalive_timeout = 5 -> null
              - origin_protocol_policy   = "https-only" -> null
              - origin_read_timeout      = 30 -> null
              - origin_ssl_protocols     = [
                  - "TLSv1.2",
                ] -> null
            }
        }
      - origin {
          - domain_name = "d1h9gz.........cloudfront.net" -> null
          - origin_id   = "nic-test - Frontend Assets" -> null

          - custom_origin_config {
              - http_port                = 80 -> null
              - https_port               = 443 -> null
              - origin_keepalive_timeout = 5 -> null
              - origin_protocol_policy   = "https-only" -> null
              - origin_read_timeout      = 30 -> null
              - origin_ssl_protocols     = [
                  - "TLSv1.2",
                ] -> null
            }
        }
      + origin {
          + domain_name = "d1h9gz.........cloudfront.net"
          + origin_id   = "nic-test - Frontend Assets"

          + custom_origin_config {
              + http_port                = 80
              + https_port               = 443
              + origin_keepalive_timeout = 5
              + origin_protocol_policy   = "https-only"
              + origin_read_timeout      = 30
              + origin_ssl_protocols     = [
                  + "TLSv1.2",
                ]
            }
        }
      + origin {
          + domain_name = (known after apply)
          + origin_id   = "nic-test - Backend API"

          + custom_origin_config {
              + http_port                = 80
              + https_port               = 443
              + origin_keepalive_timeout = 5
              + origin_protocol_policy   = "https-only"
              + origin_read_timeout      = 30
              + origin_ssl_protocols     = [
                  + "TLSv1.2",
                ]
            }
        }
```

And this is still causing the inconsistent final plan. The gateway deployment is destroyed and recreated, but this is causing the invoke_url to become unknown, even though it will be the same as it currently is.

I'm still not sure why it's modifying the frontend origins. Maybe because the order changed?

**Comment by jgou** [comment]
I have the same problem updating my Cloudfront distribution to add a lambda@edge to some of the origins. I don't know why Terraform planned to recreate the non-related origins

      - origin {
          - domain_name = "XXXXX.us-east-1.elb.amazonaws.com" -> null
          - origin_id   = "ELB-XXXXX" -> null

          - custom_origin_config {
              - http_port                = 80 -> null
              - https_port               = 443 -> null
              - origin_keepalive_timeout = 5 -> null
              - origin_protocol_policy   = "https-only" -> null
              - origin_read_timeout      = 30 -> null
              - origin_ssl_protocols     = [
                  - "SSLv3",
                  - "TLSv1",
                  - "TLSv1.1",
                  - "TLSv1.2",
                ] -> null
            }
        }
      + origin {
          + domain_name = "XXXXX.us-east-1.elb.amazonaws.com"
          + origin_id   = "ELB-XXXXX"

          + custom_origin_config {
              + http_port                = 80
              + https_port               = 443
              + origin_keepalive_timeout = 5
              + origin_protocol_policy   = "https-only"
              + origin_read_timeout      = 30
              + origin_ssl_protocols     = [
                  + "SSLv3",
                  + "TLSv1",
                  + "TLSv1.1",
                  + "TLSv1.2",
                ]
            }
        }

Then I get the following error


```
Error: Provider produced inconsistent final plan

When expanding the plan for aws_cloudfront_distribution.distribution to
include new values learned so far during apply, provider
"registry.terraform.io/-/aws" produced an invalid new value for .origin:
planned set element
cty.ObjectVal(map[string]cty.Value{"custom_header":cty.SetValEmpty(cty.Object(map[string]cty.Type{"name":cty.String,
"value":cty.String})),
"custom_origin_config":cty.ListVal([]cty.Value{cty.ObjectVal(map[string]cty.Value{"http_port":cty.NumberIntVal(80),
"https_port":cty.NumberIntVal(443),
"origin_keepalive_timeout":cty.NumberIntVal(5),
"origin_protocol_policy":cty.StringVal("https-only"),
"origin_read_timeout":cty.NumberIntVal(30),
"origin_ssl_protocols":cty.SetVal([]cty.Value{cty.StringVal("SSLv3"),
cty.StringVal("TLSv1"), cty.StringVal("TLSv1.1"),
cty.StringVal("TLSv1.2")})})}),
"domain_name":cty.StringVal("XXXXX.us-east-1.elb.amazonaws.com"),
"origin_id":cty.StringVal("ELB-XXXXX"),
"origin_path":cty.NullVal(cty.String),
"s3_origin_config":cty.ListValEmpty(cty.Object(map[string]cty.Type{"origin_access_identity":cty.String}))})
does not correlate with any element in actual.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.

```

What it is strange is that the error is related with an origin not related with Lambda@Edges. I think this is related with any updates in Cloudfront distribution because it happens frequently

The workaround for that is to destroy and create again but this not something that can be done in Production.

Is it really an issue with Terraform provider?

**Comment by c4m4** [comment]
I have this issue too:

include new values learned so far during apply, provider
"registry.terraform.io/-/null" produced an invalid new value for
.triggers["md5"]: was cty.StringVal("d5d64003774e5023c73203145ffcc3b3"), but
now cty.StringVal("24201183ca0d1b7a41bef4776fab1ba6").

This is a bug in the provider, which should be reported in the provider's own
issue tracker.

include new values learned so far during apply, provider
"registry.terraform.io/-/null" produced an invalid new value for
.triggers["md5"]: was cty.StringVal("d5d64003774e5023c73203145ffcc3b3"), but
now cty.StringVal("24201183ca0d1b7a41bef4776fab1ba6").

This is a bug in the provider, which should be reported in the provider's own
issue tracker.

include new values learned so far during apply, provider
"registry.terraform.io/-/null" produced an invalid new value for
.triggers["md5"]: was cty.StringVal("d5d64003774e5023c73203145ffcc3b3"), but
now cty.StringVal("24201183ca0d1b7a41bef4776fab1ba6").

This is a bug in the provider, which should be reported in the provider's own
issue tracker.

I have a normal null_resource as follow:

resource null_resource install_istio {
  triggers = {
    host = var.host
    md5  = filemd5(local_file.istio_profile.filename)
  }

**Comment by bflad** [comment]
Hi folks 👋 

Relating to the OpenAPI behaviors in the `aws_api_gateway_rest_api` resource, most of these should be resolved with version 3.25.0 of the Terraform AWS Provider released yesterday. Any resource configuration values will now override the relevant OpenAPI specification that was imported.

For the (re)deployment issues, this is due to Terraform configurations containing an `aws_api_gateway_deployment` resource using the `stage_name` argument. Behind the scenes when using the `stage_name` argument in the deployment resource, it is deleting and recreating the stage during resource recreation, which is why the attributes become "unknown" to downstream references such as those in the `aws_cloudfront_distribution` resource `origin`.

To fix this issue in your configuration, you should:

* Use the `aws_api_gateway_stage` resource instead of `aws_api_gateway_deployment` resource `stage_name` argument
* Ensure the `aws_api_gateway_deployment` resource has `create_before_destroy` enabled

That should stabilize the Terraform plan across API updates.

To summarize these in a quick configuration example:

```
resource "aws_api_gateway_deployment" "example" {
  rest_api_id = aws_api_gateway_rest_api.example.id

  triggers = {
    redeployment = # ... references to upstream resources (see issue link below) ...
  }

  lifecycle {
    create_before_destroy = true
  }
}

resource "aws_api_gateway_stage" "example" {
  deployment_id = aws_api_gateway_deployment.example.id
  rest_api_id   = aws_api_gateway_rest_api.example.id
  stage_name    = "example"
}

resource "aws_cloudfront_distribution" "example" {
  # ... other configuration ...

  origin {
    # ... other configuration ...
    domain_name  = # ... reference to aws_api_gateway_stage.example.invoke_url ...
  }
}
```

* For a longer explanation with more guidance and example configurations: https://github.com/hashicorp/terraform-provider-aws/issues/11344#issuecomment-765138213
* Updates for the Terraform AWS Provider's API Gateway REST API documentation: https://github.com/hashicorp/terraform-provider-aws/pull/17230

Since the above issue and pull request should exhaustively cover this problem as well, I'm going to close this issue. If you are having further trouble with this functionality after applying these recommendations, please submit new GitHub issues following the [bug report](https://github.com/terraform-providers/terraform-provider-aws/issues/new?template=Bug_Report.md) and [feature request](https://github.com/terraform-providers/terraform-provider-aws/issues/new?labels=enhancement&template=Feature_Request.md) issue templates for further triage.
